File: extensions/vscode/src/downloadDiagnostic.ts
import { randomUUID } from "node:crypto";
import { ApiError, RequestCancelledError } from "./api";

export type DownloadStage = "preflight" | "requesting" | "receiving" | "verifying" | "installing" | "files_ready" | "opening";
export class DownloadFailure extends Error {
  constructor(public readonly code: string) { super(code); }
}
export const stageLabels: Record<DownloadStage, string> = {
  preflight: "저장 위치 확인", requesting: "서버 요청·파일 수신", receiving: "파일 수신",
  verifying: "파일 무결성 검증", installing: "파일 저장", files_ready: "파일 준비 완료", opening: "IDE 폴더 열기",
};
export function classifyDownloadError(error: unknown, stage: DownloadStage, cancelled = false): string {
  if (cancelled || error instanceof RequestCancelledError) return "AG-DL-USER-CANCELLED";
  if (stage === "opening") return "AG-DL-OPEN-WORKSPACE";
  if (error instanceof DownloadFailure) return error.code;
  if (error instanceof ApiError) {
    if (error.status === 401) return "AG-DL-AUTH-EXPIRED";
    if (error.status === 403) return "AG-DL-ACCESS-DENIED";
    if (error.status === 429) return "AG-DL-HTTP-RATE-LIMIT";
    if (error.status >= 500) return "AG-DL-HTTP-SERVER";
    if (error.code === "request_timeout") return "AG-DL-NETWORK-TIMEOUT";
    if (error.code === "invalid_response") return "AG-DL-RESPONSE-TYPE";
    if (error.code === "response_too_large") return "AG-DL-INTEGRITY-SIZE";
    if (error.code === "network_error") return "AG-DL-NETWORK-UNKNOWN";
  }
  const native = error && typeof error === "object" && "code" in error ? String(error.code) : "";
  const known: Record<string,string> = { EACCES:"LOCAL-PERMISSION", EPERM:"LOCAL-PERMISSION", ENOSPC:"LOCAL-SPACE",
    EEXIST:"LOCAL-EXISTS", ENAMETOOLONG:"LOCAL-PATH", ENOENT:"LOCAL-PATH", EIO:"LOCAL-IO" };
  return "AG-DL-" + (known[native] ?? "UNKNOWN");
}
export function downloadGuidance(code?: string): string {
  switch (code) {
    case "AG-DL-LOCAL-PERMISSION": return "이 위치에 저장할 권한이 없습니다. 쓰기 가능한 다른 폴더를 선택하세요. 다시 수락할 필요는 없습니다.";
    case "AG-DL-LOCAL-SPACE": return "디스크 여유 공간을 확보한 뒤 다시 받으세요. 기존 코드는 보존하세요.";
    case "AG-DL-LOCAL-EXISTS": return "같은 이름의 폴더가 있습니다. 기존 과제를 열거나 다른 위치를 선택하세요. 기존 코드는 덮어쓰지 않습니다.";
    case "AG-DL-AUTH-EXPIRED": return "로그인이 만료됐거나 바뀌었습니다. 새 수령 코드로 로그인하세요. 기존 파일은 남습니다.";
    case "AG-DL-OPEN-WORKSPACE": return "다운로드는 완료됐습니다. 재다운로드하지 말고 현재 수업 폴더의 과제 파일을 여세요.";
    case "AG-DL-USER-CANCELLED": return "작업을 취소했습니다. 필요할 때 다시 시도하세요.";
    case "AG-DL-INTEGRITY-HASH": case "AG-DL-INTEGRITY-SIZE": return "받은 파일과 과제 정보가 다릅니다. 과제 새로고침 후 재시도하거나 문의번호를 알려 주세요.";
    case "AG-DL-RESPONSE-TYPE": return "과제 파일 대신 다른 응답을 받았습니다. API 주소와 서버 프록시 설정을 확인하세요.";
    case "AG-DL-ACCESS-DENIED": return "교수자에게 수강 상태와 과제 접근 권한을 확인해 달라고 요청하세요.";
    case "AG-DL-NETWORK-TIMEOUT": case "AG-DL-NETWORK-UNKNOWN": return "API 주소와 네트워크 연결을 확인하고 다시 시도하세요. 인증서 검증을 해제하지 마세요.";
    case "AG-DL-HTTP-RATE-LIMIT": case "AG-DL-HTTP-SERVER": return "서버가 요청을 처리하지 못했습니다. 잠시 후 다시 시도하세요.";
    case undefined: return "수락·파일 준비·IDE 열기는 별도 단계입니다.";
    default: return "실패 단계와 문의번호를 교수자에게 알려 주세요. 기존 코드는 덮어쓰지 마세요.";
  }
}
export class DownloadDiagnostic {
  readonly attemptId = randomUUID();
  stage: DownloadStage = "preflight";
  outcome = "in_progress";
  openOutcome = "not_attempted";
  code?: string;
  httpStatus?: number;
  delivery = "전달 전";
  private sequence = 0;
  fail(error: unknown, cancelled = false): void {
    this.code = classifyDownloadError(error, this.stage, cancelled);
    if (error instanceof ApiError && error.status >= 100 && error.status <= 599) this.httpStatus = error.status;
    if (this.stage === "opening") { this.outcome = "succeeded"; this.openOutcome = this.code === "AG-DL-USER-CANCELLED" ? "open_cancelled" : "open_failed"; }
    else this.outcome = this.code === "AG-DL-USER-CANCELLED" ? "cancelled" : "failed";
  }
  payload(version: string, platform: string, remote?: string): object {
    return { schema_version:1, attempt_id:this.attemptId, seq:this.sequence++, stage:this.stage,
      outcome:this.outcome, open_outcome:this.openOutcome, error_code:this.code, http_status:this.httpStatus,
      ide:"vscode", extension_version:version, os:platform === "win32" ? "windows" : platform === "darwin" ? "macos" : platform === "linux" ? "linux" : "other",
      remote_kind:!remote ? "none" : remote === "wsl" ? "wsl" : remote === "ssh-remote" ? "ssh" : "other" };
  }
  get summary(): string {
    return this.outcome === "succeeded" ? "파일 준비 완료" + (this.openOutcome === "open_failed" ? " · IDE 열기 실패" : "") : this.outcome === "cancelled" ? "다운로드 취소" : this.outcome === "failed" ? "다운로드 실패" : "다운로드 진행 중";
  }
  get text(): string {
    return `${this.summary}\n단계: ${stageLabels[this.stage]}\n오류 코드: ${this.code ?? "없음"}\n문의번호: ${this.attemptId}\nHTTP: ${this.httpStatus ?? "응답 없음/해당 없음"}\n진단: ${this.delivery}\n\n${downloadGuidance(this.code)}`;
  }
}

File: extensions/visualstudio/Autograde.Core/DownloadDiagnostic.cs
using System;
using System.IO;
using System.Net.Http;
using System.Security.Authentication;
using Newtonsoft.Json.Linq;

namespace Autograde.Core
{
    public sealed class DownloadFailure : Exception
    {
        public string Code { get; }
        public DownloadFailure(string code) : base(code) { Code = code; }
    }

    // An in-memory, allowlisted support record. Never serialize exception messages or paths.
    public sealed class DownloadDiagnostic
    {
        public string AttemptId { get; } = Guid.NewGuid().ToString();
        public string Stage { get; set; } = "preflight";
        public string Outcome { get; set; } = "in_progress";
        public string OpenOutcome { get; set; } = "not_attempted";
        public string Code { get; private set; }
        public int? Status { get; private set; }
        public string Delivery { get; set; } = "전달 전";
        int sequence;

        public void Fail(Exception error, bool cancelled)
        {
            Code = Classify(error, cancelled, Stage);
            if (error is ServiceError service) Status = service.Status;
            if (Stage == "opening") { Outcome = "succeeded"; OpenOutcome = cancelled ? "open_cancelled" : "open_failed"; }
            else Outcome = cancelled ? "cancelled" : "failed";
        }
        public void MarkOpened()
        {
            Stage = "opening"; Outcome = "succeeded"; OpenOutcome = "opened";
            Code = null; Status = null;
        }
        // Local UI only: never append this path-bearing text to Payload or Details.
        public static string OpenRecoveryGuidance(string target) =>
            "다운로드한 파일은 보존됩니다. 다시 수락하거나 재다운로드할 필요는 없습니다.\n" +
            "저장 위치: " + target + "\n" +
            "1. 기존 작업을 모두 저장(Ctrl+Shift+S)하고 저장·신뢰 확인창이 열려 있는지 확인하세요. 신뢰할 수 있는 수업 자료만 여세요.\n" +
            "2. ‘과제 폴더 열기’를 다시 누르세요.\n" +
            "3. 계속 실패하면 Visual Studio의 파일 → 열기 → 폴더에서 위 저장 위치를 선택하세요. 상위 폴더가 아닌 main.c/main.cpp 또는 README가 있는 과제 폴더입니다.\n" +
            "4. ‘폴더 관리 · 다시 다운로드’의 제출 폴더가 같은 위치인지 확인하고, 수정 후 모두 저장하여 제출하세요.\n" +
            "계속 열리지 않으면 ‘다운로드 상세 · 문의번호’의 문의번호·오류 코드와 VS 버전을 교수자에게 전달하세요. 비밀번호·수령 코드는 보내지 마세요.";
        public JObject Payload(string version) => new JObject {
            ["schema_version"] = 1, ["attempt_id"] = AttemptId, ["seq"] = sequence++,
            ["stage"] = Stage, ["outcome"] = Outcome, ["open_outcome"] = OpenOutcome,
            ["error_code"] = Code, ["http_status"] = Status,
            ["ide"] = "visualstudio", ["extension_version"] = version,
            ["os"] = "windows", ["remote_kind"] = "none"
        };
        public string Summary => Outcome == "succeeded" ? (OpenOutcome == "open_failed" ? "파일 준비 완료 · IDE 열기 실패" : OpenOutcome == "open_cancelled" ? "파일 준비 완료 · 열기 취소" : "과제 파일 준비 완료") : Outcome == "cancelled" ? "다운로드 취소 · 수락은 유지됩니다." : "다운로드 " + (Outcome == "failed" ? "실패" : "진행 중") + " · " + StageLabel(Stage);
        public string Details => Summary + "\n단계: " + StageLabel(Stage) + "\n오류 코드: " + (Code ?? "없음") +
            (Status.HasValue ? "\nHTTP: " + Status : "") + "\n문의번호: " + AttemptId + "\n진단: " + Delivery + "\n" + Guidance(Code);
        public static string StageLabel(string stage)
        {
            switch (stage) { case "preflight": return "저장 위치 확인"; case "requesting": return "서버 요청·파일 수신";
                case "verifying": return "파일 무결성 검증"; case "installing": return "파일 저장";
                case "files_ready": return "파일 준비 완료"; case "opening": return "IDE 폴더 열기"; default: return "파일 수신"; }
        }
        public static string Classify(Exception error, bool cancelled, string stage)
        {
            string code;
            if (cancelled) code = "USER-CANCELLED";
            else if (stage == "opening") code = "OPEN-WORKSPACE";
            else if (error is DownloadFailure typed) return typed.Code;
            else if (error is ServiceError service) code = service.Status == 401 ? "AUTH-EXPIRED" : service.Status == 403 ? "ACCESS-DENIED" : service.Status == 429 ? "HTTP-RATE-LIMIT" : service.Status >= 500 ? "HTTP-SERVER" : "UNKNOWN";
            else if (error is OperationCanceledException) code = "NETWORK-TIMEOUT";
            else if (error is UnauthorizedAccessException) code = "LOCAL-PERMISSION";
            else if (error is PathTooLongException) code = "LOCAL-PATH";
            else if (error is AuthenticationException || error.InnerException is AuthenticationException) code = "NETWORK-TLS";
            else if (error is HttpRequestException) code = "NETWORK-UNKNOWN";
            else if (error is InvalidDataException) code = "ARCHIVE-INVALID";
            else if (error is IOException io) {
                int native = io.HResult & 0xffff;
                code = native == 112 || native == 39 ? "LOCAL-SPACE" : native == 5 ? "LOCAL-PERMISSION" : native == 80 || native == 183 ? "LOCAL-EXISTS" : "LOCAL-IO";
            }
            else code = "UNKNOWN";
            return "AG-DL-" + code;
        }
        public static string Guidance(string code)
        {
            switch (code) {
                case "AG-DL-LOCAL-PERMISSION": return "선택한 폴더에 저장할 권한이 없습니다. 다른 저장 폴더를 선택하세요. 다시 수락할 필요는 없습니다.";
                case "AG-DL-LOCAL-SPACE": return "디스크 공간을 확보한 뒤 다시 받으세요. 기존 코드는 삭제하지 않습니다.";
                case "AG-DL-LOCAL-EXISTS": return "기존 폴더를 열거나 새 폴더에 받으세요. 기존 코드는 덮어쓰지 않습니다.";
                case "AG-DL-AUTH-EXPIRED": return "로그인이 만료되었습니다. 새 수령 코드로 로그인하세요. 기존 파일은 보존됩니다.";
                case "AG-DL-ACCESS-DENIED": return "과제 접근이 거부됐습니다. 교수자에게 수강·과제 권한 확인을 요청하세요.";
                case "AG-DL-NETWORK-TLS": return "도메인과 인증서를 확인하세요. 인증서 검증을 해제하지 마세요.";
                case "AG-DL-OPEN-WORKSPACE": return "다운로드 완료 · IDE 자동 열기만 실패했습니다. 기존 작업 저장 후 ‘과제 폴더 열기’로 재시도하거나 파일 → 열기 → 폴더에서 저장 위치를 여세요. 아래 안내에서 경로와 복구 순서를 확인하세요.";
                case "AG-DL-USER-CANCELLED": return "취소했습니다. 필요할 때 다시 시도하세요.";
                case "AG-DL-INTEGRITY-HASH": case "AG-DL-INTEGRITY-SIZE": return "받은 파일과 과제 정보가 다릅니다. 과제 정보를 새로고침하고 재시도하거나 문의번호를 알려 주세요.";
                case "AG-DL-RESPONSE-TYPE": return "과제 파일 대신 다른 응답을 받았습니다. API 주소와 서버 프록시 설정을 확인하세요.";
                case "AG-DL-HTTP-RATE-LIMIT": case "AG-DL-HTTP-SERVER": return "서버가 요청을 처리하지 못했습니다. 잠시 후 다시 시도하세요.";
                case "AG-DL-NETWORK-TIMEOUT": case "AG-DL-NETWORK-UNKNOWN": return "서버 연결·응답을 확인할 수 없습니다. API 주소와 네트워크를 확인하고 재시도하세요.";
                case null: return "수락과 다운로드·IDE 열기는 별도 단계입니다.";
                default: return "문의번호와 실패 단계를 교수자에게 알려 주세요. 기존 코드는 덮어쓰지 마세요.";
            }
        }
    }
}
